File: LabFinal/5_fifo.py
n = int(input("Enter number of pages: "))
pages = list(map(int, input("Enter page reference string: ").split()))
f = int(input("Enter number of frames: "))

frames = []
for i in range(f):
    frames.append(-1)

miss = 0
hit = 0
idx = 0 

for i in range(n):
    found = 0

    for j in range(f):
        if frames[j] == pages[i]:
            found = 1
            hit += 1
            break

    if found == 0:
        # MISS
        frames[idx] = pages[i]
        idx = (idx + 1) % f
        miss += 1
        print("Page", pages[i], "-> FAULT ->", end=" ")
    else:
        print("Page", pages[i], "-> HIT   ->", end=" ")

    # frame
    for j in range(f):
        print(frames[j], end=" ")

    print()

print("\nPage Miss =", miss)
print("Page Hits =", hit)

File: LabFinal/5_fifo.cpp
#include <iostream>
using namespace std;

const int MAX = 50;

int main() {

    int n, f;

    int pages[MAX];
    int frames[MAX];

    int miss = 0;
    int hit = 0;

    int idx = 0;

    // Input number of pages
    cout << "Enter number of pages: ";
    cin >> n;

    // Input page reference string
    cout << "Enter page reference string: ";

    for (int i = 0; i < n; i++) {
        cin >> pages[i];
    }

    // Input number of frames
    cout << "Enter number of frames: ";
    cin >> f;

    // Initialize frames with -1
    for (int i = 0; i < f; i++) {
        frames[i] = -1;
    }

    // FIFO Page Replacement
    for (int i = 0; i < n; i++) {

        bool found = false;

        // Check HIT
        for (int j = 0; j < f; j++) {

            if (frames[j] == pages[i]) {
                found = true;
                hit++;
                break;
            }
        }

        // MISS
        if (!found) {

            frames[idx] = pages[i];

            // Circular index
            idx = (idx + 1) % f;

            miss++;

            cout << "Page " << pages[i]
                 << " -> FAULT -> ";
        }
        else {

            cout << "Page " << pages[i]
                 << " -> HIT   -> ";
        }

        // Print frames
        for (int j = 0; j < f; j++) {
            cout << frames[j] << " ";
        }

        cout << endl;
    }

    // Output
    cout << "\nPage Miss = " << miss << endl;
    cout << "Page Hits = " << hit << endl;

    return 0;
}